krate/src/main/kotlin/hu/autsoft/krate/Functions.kt

```diff
@@ -100,3 +105,123 @@ public fun Krate.stringPref(key: String, defaultValue: String): ReadWritePropert
+    return ValidatedPreferenceDelegate(IntDelegateWithDefault(key, default), isValid)
+}
+
+/**
+ * Creates a validated optional preference of type [String] with a given [key]
+ *
+ * @param [key] the key to identify this value in SharedPreferences
+ * @param [isValid] the function to validate values during set
+ *
+ * @return [ReadWriteProperty]
+ */
+public fun Krate.stringPref(key: String, isValid: (String?) -> Boolean): ReadWriteProperty<Krate, String?> {
+    return ValidatedPreferenceDelegate(StringDelegate(key), isValid)
+}
+
+/**
+ * Creates a validated non-optional preference of type [String] with a given [key]
+ *
+ * @param [key] the key to identify this value in SharedPreferences
+ * @param [isValid] the function to validate values during set
+ *
+ * @return [ReadWriteProperty]
+ */
+public fun Krate.stringPref(key: String, default: String, isValid: (String) -> Boolean): ReadWriteProperty<Krate, String> {
+    return ValidatedPreferenceDelegate(StringDelegateWithDefault(key, default), isValid)
+}
+
+/**
+ * Creates a validated optional preference of type [Set] with a given [key]
+ *
+ * @param [key] the key to identify this value in SharedPreferences
+ * @param [isValid] the function to validate values during set
+ *
+ * @return [ReadWriteProperty]
+ */
+public fun Krate.stringSetPref(key: String, isValid: (Set<String>?) -> Boolean): ReadWriteProperty<Krate, Set<String>?> {
+    return ValidatedPreferenceDelegate(StringSetDelegate(key), isValid)
+}
+
+/**
+ * Creates a validated non-optional preference of type [Set] with a given [key]
+ *
+ * @param [key] the key to identify this value in SharedPreferences
+ * @param [isValid] the function to validate values during set
+ *
+ * @return [ReadWriteProperty]
+ */
+public fun Krate.stringSetPref(key: String, defaultValue: Set<String>, isValid: (Set<String>) -> Boolean): ReadWriteProperty<Krate, Set<String>> {
+    return ValidatedPreferenceDelegate(StringSetDelegateWithDefault(key, defaultValue), isValid)
+}
+
+/**
+ * Creates a validated optional preference of type [Float] with a given [key]
+ *
+ * @param [key] the key to identify this value in SharedPreferences
+ * @param [isValid] the function to validate values during set
+ *
+ * @return [ReadWriteProperty]
+ */
+public fun Krate.floatPref(key: String, isValid: (Float?) -> Boolean): ReadWriteProperty<Krate, Float?> {
+    return ValidatedPreferenceDelegate(FloatDelegate(key), isValid)
+}
+
+/**
+ * Creates a validated non-optional preference of type [Float] with a given [key]
+ *
+ * @param [key] the key to identify this value in SharedPreferences
+ * @param [isValid] the function to validate values during set
+ *
+ * @return [ReadWriteProperty]
+ */
+public fun Krate.floatPref(key: String, default: Float, isValid: (Float) -> Boolean): ReadWriteProperty<Krate, Float> {
+    return ValidatedPreferenceDelegate(FloatDelegateWithDefault(key, default), isValid)
+}
```

Comment:
The functions originally defined in this file had the order of optionals first (alphabetically: boolean, float, int, long, string, stringset) and then non-optionals. Would be nice if the validated functions had the same order.

Also, organization-wise, the validated functions could go in a different file (`ValidatedFunctions.kt`?).

I've also thought about having them in their own package, but being in the same package will help moving between overloads for these functions without having to change imports, so I think they should stay in `hu.autsoft.krate` as they are now.

Reply:
How should we do this for the gson module? I attempted to add the overloads there, but there was something weird with the way the generics were being done. 

```kotlin
public inline fun <reified T : Any> Krate.gsonPref(key: String): ReadWriteProperty<Krate, T?> {
    return gsonPrefImpl(key, object : TypeToken<T>() {}.type)
}
```

Those above code's type signature `fun <reified T : Any>` I've always seen used to specify that the generic must _not_ be nullable. Seems to me we should drop the `T : Any` and allow the user to specify if their type should be nullable or not.

Reply:
Does `TypeToken` not take into account nullability? If so, maybe it makes sense to look into `kolinx-serialization`?

Reply:
Indeed, `TypeToken` doesn't care about nullability - Gson has no notion of nullability. So it doesn't matter whether we create it passing in `T` or `T?` here.

This doesn't really matter in practice, because the behaviour is correct. Making a `fromJson` call with this `TypeToken` will return `null` if there's no `String` to parse.

Reply:
Of course, for even nicer type safety, a Moshi or kotlinx.serialization module could be added. Gson was just a widely used library that we've decided to add additional support for first.